endpoints/openrtb2/auction.go

```diff
@@ -350,135 +350,143 @@ func parseTimeout(requestJson []byte, defaultTimeout time.Duration) time.Duratio
+		return nil
+	}
 
-		if len(imp.Ext) == 0 {
-			imps = append(imps, updatedImp)
-			continue
-		}
+	bidderParamsJson := prebid.BidderParams
+	if len(bidderParamsJson) == 0 {
+		return nil
+	}
 
-		var impExt map[string]json.RawMessage
-		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
-			// imp.ext should be parsable, but we haven't validated yet so there may be another issue. ignore it here
-			// and let the validation catch it later on.
-			imps = append(imps, updatedImp)
-			continue
-		}
+	bidderParams := map[string]map[string]json.RawMessage{}
+	if err := json.Unmarshal(bidderParamsJson, &bidderParams); err != nil {
+		return nil
+	}
 
-		//merges bidder parameters passed at req.ext level with imp[].ext level.
-		impExtModified, err := addMissingReqExtParamsInImpExt(impExt, reqBidderParams)
+	for i, imp := range req.GetImp() {
+		impExt, err := imp.GetImpExt()
 		if err != nil {
-			return fmt.Errorf("error processing bidder parameters for imp[%d]: %s", impIndex, err.Error())
+			continue
 		}
 
-		//merges bidder parameters passed at req.ext level with imp[].ext.prebid.bidder level.
-		impExtPrebidModified, err := addMissingReqExtParamsInImpExtPrebid(impExt, reqBidderParams)
-		if err != nil {
-			return fmt.Errorf("error processing bidder parameters for imp[%d]: %s", impIndex, err.Error())
+		// merges bidder parameters passed at req.ext level with imp[].ext.BIDDER level
+		if err := mergeBidderParamsImpExt(impExt, bidderParams); err != nil {
+			return fmt.Errorf("error processing bidder parameters for imp[%d]: %s", i, err.Error())
 		}
 
-		if impExtModified || impExtPrebidModified {
-			updatedImpExt, err := json.Marshal(impExt)
-			if err != nil {
-				return fmt.Errorf("error processing bidder parameters for imp[%d]: error marshalling ext:%s", impIndex, err.Error())
-			}
-			updatedImp.Ext = updatedImpExt
+		// merges bidder parameters passed at req.ext level with imp[].ext.prebid.bidder.BIDDER level
+		if err := mergeBidderParamsImpExtPrebid(impExt, bidderParams); err != nil {
+			return fmt.Errorf("error processing bidder parameters for imp[%d]: %s", i, err.Error())
 		}
-
-		imps = append(imps, updatedImp)
 	}
 
-	req.BidRequest.Imp = imps
 	return nil
 }
 
-// addMissingReqExtParamsInImpExtPrebid merges bidder parameters passed at req.ext level with imp[].ext.prebid.bidder level.
-func addMissingReqExtParamsInImpExtPrebid(impExtByKey map[string]json.RawMessage, reqExtParams map[string]map[string]json.RawMessage) (bool, error) {
-	if _, prebidSectionExists := impExtByKey["prebid"]; !prebidSectionExists {
-		return false, nil
-	}
+// mergeBidderParamsImpExt merges bidder parameters in req.ext down to the imp[].ext.BIDDER
+// level, giving priority to imp[].ext.BIDDER in case of a conflict. Unmarshal errors are not
+// expected since the ext json was validated during the bid request unmarshal.
+func mergeBidderParamsImpExt(impExt *openrtb_ext.ImpExt, reqExtParams map[string]map[string]json.RawMessage) error {
```

Comment:
Still need this for now because bidder param resolution happens before validation where the bidder params are moved to imp[].ext.prebid.bidder.BIDDER. Leaving to be addressed in future work. This PR is complex enough.